Disconnect all slots on Aktualizr destructor

patriotyk opened the pull request:
This PR fixes APICheck test fails. 
But the problem was in Aktualizr class where we didn't disconnects from connected slots at destruction.

Signed-off-by: Serhiy Stetskovych <patriotyk@gmail.com>

OYTIS commented on src/libaktualizr/primary/aktualizr_test.cc:
```diff
@@ -926,13 +926,14 @@ TEST(Aktualizr, APICheck) {
     boost::signals2::connection conn = aktualizr.SetSignalHandler(f_cb);
     aktualizr.Initialize();
-    for (int i = 0; i < 100; ++i) {
+    for (int i = 0; i < 101; ++i) {
```
Why?

patriotyk replied:
To make 100 cycles, otherwise we make only 99

OYTIS replied:
Um... no.

patriotyk replied:
I can change it to 'i <=100 '

eu-smirnov replied:
There are 100 numbers in the range [0, 99]. `i < 100` is exactly what we want.

eu-smirnov replied:
@patriotyk happens to me as well sometimes :)

OYTIS commented:
Solves the problem for me. I'm not completely comfortable with the solution though. We should clearly document that our users are never supposed to run multiple instances of Aktualizr. Or maybe `shared_ptr` is just wrong abstraction here.
Nevertheless, I believe it should be merged.

eu-smirnov commented on src/libaktualizr/primary/aktualizr_test.cc:
```diff
@@ -944,6 +945,14 @@ class HttpPutManifestFail : public HttpFake {
+void process_events_PutManifestError(const std::shared_ptr<event::BaseEvent>& event) {
+  std::cout << event->variant << "\n";
+  if (event->variant == "UpdateCheckComplete") {
+    num_events_PutManifestError++;
```
We should check here, that it's really an error.

patriotyk replied:
We do that in 947 line

eu-smirnov replied:
Well, there we check a result, not an event. I think it won't hurt to check if event status matches the result.

patriotyk replied:
OK, Updated

patrickvacek commented:
> We should clearly document that our users are never supposed to run multiple instances of Aktualizr at least.

You are probably right that we should make that clear. At first glance, it seems obvious to me, but perhaps it isn't. Even better: is there a way we could detect if aktualizr is already running and abort?

> Or maybe `shared_ptr` is just wrong abstraction here.

For what exactly?

OYTIS commented:
@patrickvacek 
> For what exactly?

For sharing an event channel. Channel is obviously something asymmetric, while `shared_ptr` is symmetric. Moreover, `shared_ptr` makes an impression that it can be shared between multiple instances of `Aktualizr`, which it even technically can.

OYTIS commented:
Maybe `events::Channel` should not be really exposed to Aktualizr user. It was convenient when a channel was a queue of events, but now that we are using `boost::signal2` with callbacks, I see little value and a decent amount of confusion here.

OYTIS commented:
Yes, you're right, the public constructor doesn't allow the user to do it. It's worth fixing the test then to not abuse the private constructor (like, keep the scope of `event::Channel` the same as the scope of `Aktualizr`). Would be better than `disconnect_all_slots` IMO.

OYTIS commented:
This is a more pleasant solution IMO: https://github.com/advancedtelematic/aktualizr/pull/1037

patriotyk commented:
Oh I just only noticed that sig_ comes from outside of Aktualizr instance and dissconnecting slots inside is wrong. So I agree with Anton's #1037 PR. But it complements my PR, not replaces.

patrickvacek commented:
> dissconnecting slots inside is wrong.

This PR is still disconnecting the slots, though. Is that unintentional now? I think I agree with Anton that i'd rather not mess with the slots internally. It seems too likely to be hiding other problems.

patriotyk commented:
It disconnect slots in destructor but for now member is incapsulated and never exposed. So there is no any drawback to dissconnect slots from that member which will be destroyed immediatelyy after Aktualizr destructor finished.

patriotyk commented:
But If we destroy this object with destructor we should not worry about disconnection and all slots will be disconnected implicitly. It is maybe beter to remove this line of code. So I have updated PR and removed it.
